R/node-match.R

```diff
@@ -0,0 +1,454 @@
+#' Match patterns to a call
+#'
+#' @description
+#'
+#' `node_match()` is a handy way of checking whether a quoted call
+#' conforms to a pattern. The following rules apply to patterns:
+#'
+#' * Named arguments. They will match even if not in order. The
+#'   patterns `call(foo = 1, bar = 2)` and `call(bar = 2, foo = 1)`
+#'   are thus equivalent.
+#'
+#' * Unnamed arguments. They will match only in order. The patterns
+#'   `call(foo, bar)` and `call(bar, foo)` are thus not equivalent.
+#'
+#' * NULL-named arguments. A NULL name signals an unnamed argument
+#'   that should match even if not in order. The patterns `call(NULL =
+#'   foo, NULL = bar)` and `call(NULL = bar, NULL = foo)` will both
+#'   match the call `call(foo, bar)`.
+#'
+#'
+#' In addition, patterns can contain wildcards that will match
+#' anything:
+#'
+#' * The `.` wildcard will match any argument. It can appear on the
+#'   LHS or the RHS of an argument. For instance `call(. = .)` will
+#'   match any argument no matter its name.
+#'
+#' * The `...` wildcard matches all remaining unmatched arguments.
```

Comment:
Maybe emphasise that the default matching is exhaustive? `i.e. f(x = 1, y = 1)` won't match `f(x = .)`